File: advent-of-code/2015/01/02/solution.cs
static int FindBasementPosition(string source)
{
    var floor = 0;
    for (var position = 0; position < source.Length; position++)
    {
        var character = source[position];
        
        if (character == '(')
            floor++;
        else if (character == ')')
            floor--;
        
        if (floor == -1)
            return position + 1; // 1-indexed
    }
    return 0;
}

// Test cases with assertions
System.Diagnostics.Debug.Assert(FindBasementPosition(")") == 1);
System.Diagnostics.Debug.Assert(FindBasementPosition("()())") == 5);

var scriptDir = Directory.GetCurrentDirectory();
var inputPath = Path.Combine(scriptDir, "..", "input.txt");
Console.WriteLine($"C#: {FindBasementPosition(File.ReadAllText(inputPath))}");

File: advent-of-code/2015/01/02/solution.rs
use std::fs;

fn find_basement_position(source: &str) -> usize {
    let mut floor = 0;
    for (position, character) in source.chars().enumerate() {
        match character {
            '(' => floor += 1,
            ')' => floor -= 1,
            _ => {}
        }

        if floor == -1 {
            return position + 1; // 1-indexed
        }
    }
    0
}

fn main() {
    // Test cases with assertions
    assert_eq!(find_basement_position(")"), 1);
    assert_eq!(find_basement_position("()())"), 5);

    // Calculate and print the answer
    let script_dir = std::path::Path::new(file!()).parent().unwrap();
    let input_path = script_dir.join("..").join("input.txt");
    let instructions = fs::read_to_string(input_path).unwrap();
    println!("Rust: {}", find_basement_position(&instructions));
}
